common/protob/messages-ethereum.proto

```diff
@@ -110,6 +110,44 @@ message EthereumMessageSignature {
     optional string address = 3;     // address used to sign the message
 }
 
+/**
+ * Request: Ask device to sign typed data
+ * @start
+ * @next EthereumTypedDataRequest
+ * @next Failure
+ */
+message EthereumSignTypedData {
+    repeated uint32 address_n = 1;  // BIP-32 path to derive the key from master node
+    optional uint32 num_members = 2; // Check if this can be removed, since it should always be 2 (domain and message/primaryType)
+}
+
+/**
+ * Response: Device asks for more values from typed data, or returns the signature and address.
+ * If member_path is not empty, device awaits the information for this value.
+ * Otherwise, the signature_* fields contain the computed transaction signature. All three fields will be present.
+ * @end
+ * @next EthereumTypedDataAck
+ * @next Failure
+ */
+message EthereumTypedDataRequest {
+    
+    repeated uint32 member_path = 1; // Path to value that should be send to device
+    optional bytes signature = 2;   // signature of the message
+    optional string address = 3;     // address used to sign the message
+}
+
+/**
+ * Request: Typed data value. 
+ * If the member is a struct member_value should not be set. If the member is literal num_members should not be set.
+ * @next EthereumTypedDataRequest
+ */
+message EthereumTypedDataAck {
+    optional string member_name = 1;
+    optional string member_type = 2;
```

Comment:
type should be an enum with allowed types.

value should prooobably be something other than a string too - maybe something like:
```protobuf
optional bytes value_bytes;
optional string value_string;
optional uint64 value_number;
```
but maybe not. i didn't come up with anything that feels "right" just yet.

Reply:
Yes that would be nicer, but that would be quite some types (around 100)

Reply:
In case it an enum with all the types is used, one way to transmit the value always as bytes (already encoded for ethereum) and just decode if necessary for display purpose.

Another pain point would be that it is required to have the string name of each type for the hashing process (it is part of the struct schema), so it would be required to have a mapping from enum type to type name in code 😕

Reply:
Instead of 100 types it would be possible to split it into a base type (e.g. int, uint, byte, address, bool, string, bytes) and width (max 32 for int, uint, byte and 0 for address, bool, string, bytes). This way it is also quite easy to rebuild the string name of the type and the enum would be manageable. For arrays that would be a mess again, but they are not officially supported yet anyways.